espnet2/enh/espnet_model.py

```diff
@@ -215,488 +132,52 @@ def forward(
 
         # for data-parallel
         speech_ref = speech_ref[:, :, : speech_lengths.max()]
-        speech_mix = speech_mix[:, : speech_lengths.max()]
-
-        loss, speech_pre, others, out_lengths, perm = self._compute_loss(
-            speech_mix,
-            speech_lengths,
-            speech_ref,
-            dereverb_speech_ref=dereverb_speech_ref,
-            noise_ref=noise_ref,
-        )
-
-        # add stats for logging
-        if self.loss_type not in ("snr", "ci_sdr", "si_snr"):
-            if self.training:
-                si_snr = None
-            else:
-                speech_pre = [self.decoder(ps, speech_lengths)[0] for ps in speech_pre]
-                speech_ref = torch.unbind(speech_ref, dim=1)
-                if speech_ref[0].dim() == 3:
-                    # For si_snr loss, only select one channel as the reference
-                    speech_ref = [sr[..., self.ref_channel] for sr in speech_ref]
```

Comment:
We should reuse this part because sometimes the reference can be single-channel while the input is multi-channel.

Reply:
Done.